---
title: "Introduction to fastbaps"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output: 
  html_document:
    fig_width: 12
    fig_height: 8
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
```

The fast BAPS algorithm is based on applying the hierarchical Bayesian clustering (BHC) algorithm of [@Heller2005-kp] to the problem of clustering genetic sequences using the same likelihood as BAPS [@Cheng2013-mp]. The Bayesian hierarchical clustering can be initiated with sequences as individual clusters or by running a faster conventional hierarchical clustering initially followed by BHC of the resulting clusters.

The algorithm has been written to take advantage of fast sparse matrix libraries and is able to handle 1000's of sequences and 100,000's of SNPs in under an hour on a laptop using a single core.

Alternatively, we can condition on an initial phylogentic or hierarchical tree and provide the partition of the hierarchy that maximises the BAPS likelihood. This is useful if the user is mainly interested in partitioning an already calculated phylogeny. We have also noticed that partitioning a hierarchy built using ward.D2 distance gives very reasonable results, very quickly.

***


## Libraries

```{r}
library(fastbaps)
library(rhierbaps)
library(ggtree)
library(phytools)
library(ggplot2)

set.seed(1234)
```

## Loading data

We first need to load a multiple sequence alignment into sparse format. We can choose between the original BAPS prior or a prior proportional to the mean frequency of each allele in the population.

```{r}
fasta.file.name <- system.file("extdata", "seqs.fa", package = "fastbaps")
sparse.data <- import_fasta_sparse_nt(fasta.file.name)
```


## Running fastbaps

It is a good idea to choose `k.init` to be significantly larger than the number of clusters you expect. By default it is set to the number of sequences / 4.

```{r}
baps.hc <- fast_baps(sparse.data)
```

This provides a Bayesian hierarchical clustering of the data. To obtain the partition of this hierarchy that maximises the marginal likelihood run

```{r}
best.partition <- best_baps_partition(sparse.data, as.phylo(baps.hc))
```

We can compare the log marginal likelihood with that obtained using hierbaps.

```{r}
snp.matrix <- load_fasta(fasta.file.name)
hb.results <- hierBAPS(snp.matrix, max.depth = 2, n.pops = 50, quiet = TRUE)

calc_marginal_llk(sparse.data, hb.results$partition.df$`level 1`)
calc_marginal_llk(sparse.data, best.partition)
```

We can also plot the output of the two algorithms along with a pre-calculated tree using ggtree [@Yu2017-bf].

```{r}
newick.file.name <- system.file("extdata", "seqs.fa.treefile", package = "fastbaps")
iqtree <- phytools::read.newick(newick.file.name)
plot.df <- data.frame(id=colnames(sparse.data$snp.matrix),
                      val=hb.results$partition.df$`level 1`,
                      fastbaps=best.partition,
                      stringsAsFactors = FALSE)

gg <- ggtree(iqtree)
gg <- gg %<+% data.frame(id=colnames(sparse.data$snp.matrix),
                      hierBAPS=hb.results$partition.df$`level 1`)
gg <- gg + geom_tippoint(aes(color = factor(hierBAPS)))

f1 <- facet_plot(gg, panel="hierBAPS", data=plot.df, geom=geom_tile, aes(x=val), color='red')
f2 <- facet_plot(f1, panel="fastbaps", data=plot.df, geom=geom_tile, aes(x=fastbaps), color='blue')
f2
```

Rather than choosing either the BAPS prior or the population means we can perform a grid search between these two priors to optimise our choice using Bayes factors.

```{r}
sparse.data <- optimise_prior(sparse.data)
baps.hc <- fast_baps(sparse.data)
best.partition <- best_baps_partition(sparse.data, as.phylo(baps.hc))
```

```{r}
plot.df <- data.frame(id=colnames(sparse.data$snp.matrix),
                      val=hb.results$partition.df$`level 1`,
                      fastbaps=best.partition,
                      stringsAsFactors = FALSE)

gg <- ggtree(iqtree)
gg <- gg %<+% data.frame(id=colnames(sparse.data$snp.matrix),
                      hierBAPS=hb.results$partition.df$`level 1`)
gg <- gg + geom_tippoint(aes(color = factor(hierBAPS)))

f1 <- facet_plot(gg, panel="hierBAPS", data=plot.df, geom=geom_tile, aes(x=val), color='red')
f2 <- facet_plot(f1, panel="fastbaps", data=plot.df, geom=geom_tile, aes(x=fastbaps), color='blue')
f2
```

##Calculating pairwise SNP similarity and distance matrices

The fastbaps package includes functions for very quickly obtaining pairwise SNP matrices. It achieves this by taking advantage of very fast sparse matrix algebra libraries. 

```{r}
snp.similarity.matrix <- snp_similarity(sparse.data)
snp.distance.matrix <- snp_dist(sparse.data)
```

By using the distance matrix to very quickly calculate a hierarchy using Ward's method we can get a reasonable result very quickly.

```{r}
d <- as.dist(snp.distance.matrix/max(snp.distance.matrix))
h <- hclust(d, method = "ward.D2")
best.partition.ward <- best_baps_partition(sparse.data, as.phylo(h))
calc_marginal_llk(sparse.data, best.partition.ward)
```


## References

---
nocite: '@*'
...

##Appendix
```{r}
sessionInfo()
```
